src/crimson/os/seastore/transaction.h

```diff
@@ -83,12 +94,37 @@ class Transaction {
     ceph_assert(inserted);
   }
 
-  void add_fresh_extent(CachedExtentRef ref) {
+  void add_fresh_extent(
+    CachedExtentRef ref,
+    delay_inline_ool_t delay_state)
+  {
     ceph_assert(!is_weak());
-    fresh_block_list.push_back(ref);
-    ref->set_paddr(make_record_relative_paddr(offset));
-    offset += ref->get_length();
-    write_set.insert(*ref);
+    switch (delay_state) {
+    case delay_inline_ool_t::NO_DELAY:
+      //XXX: this is actually left for persistent memory-like
+      //     devices, and should set absolute addresses here
+    case delay_inline_ool_t::DELAYED_INLINE:
+      ref->set_paddr(make_record_relative_paddr(offset));
+      offset += ref->get_length();
+    case delay_inline_ool_t::DELAYED_OOL:
+      // for delayed ool extents, absolute should have been set
+      assert(!ref->get_paddr().is_null());
+      if (delay_state >= delay_inline_ool_t::DELAYED_INLINE) {
```

Comment:
From this, I infer that you meant DELAYED_INLINE to fall through, but I really don't think it's worth it.  Please just duplicate the fresh_block_list and write_set calls and avoid the fallthrough.

I also don't really understand what's going on with delayed_set here.  Is the intention for add_fresh_extent t be called on the same extent twice?

Reply:
> I also don't really understand what's going on with delayed_set here. Is the intention for add_fresh_extent t be called on the same extent twice -- once as DELAYED once as DELAYED_INLINE|DELAYED_OOL?

yes, we have to keep these delayed set in the transaction, so that it can be looked up by other parts of the transaction. Am I right?

Reply:
> From this, I infer that you meant DELAYED_INLINE to fall through, but I really don't think it's worth it. Please just duplicate the fresh_block_list and write_set calls and avoid the fallthrough.

ok:-)